### PR TITLE
[FIRRTL][HW] Shared InstancePathCache between HW and FIRRTL

### PR DESCRIPTION
**Pull request by nandor**
Instance path enumeration considers the top-level node of the graph as the starting point of the enumeration.
In FIRRTL, this is the top-level of the circuit, while in HW all public modules are considered.

### CODE REVIEW

**Comment by seldridge** [comment]
Also, there's some clang-tidy issues related to `auto *foo` vs. `auto foo`. 😅

File: test/Dialect/FIRRTL/annotations.mlir [hunk 1]
```diff
@@ -58,10 +58,10 @@ firrtl.circuit "Foo" attributes {rawAnnotations = [
   // CHECK-SAME:     {class = "circt.test", data = "Target"}
   // CHECK-SAME:     {class = "circt.test", data = "ModuleName"}
   firrtl.module @Foo() {}
-  // CHECK:      firrtl.extmodule @Blackbox
+  // CHECK:      firrtl.extmodule private @Blackbox
```
[the remark this answers is not in the document]
Comment (fabianschuiki) [reply test/Dialect/FIRRTL/annotations.mlir:61]: Is the addition of `private` here really necessary for this PR? It doesn't seem to influence the annotations at all.

**Review by fabianschuiki — APPROVED** [review]
LGTM! I don't think all the `private` additions are necessary. A circuit can have multiple public modules, and you can have multiple disconnected top-level modules. I don't think it's a good idea to assume that there's only one module marked `public`, and that that is the top-level. However, I think by using the InstanceGraph, which creates this dummy root node that has dummy instances for all public modules, should still make things work. You just get a lot of superfluous instance paths. We might actually be missing a utility like `getTopModuleFor(HWModuleLike)`, which goes upwards in the instance graph to the closest root. You could then filter out paths that don't start at that root.

**Comment by nandor** [comment]
I've removed the requirement to start an enumeration from all public modules, considering instead the one toplevel for FIRRTL.

**Pull request merged**